### DISCUSSION TITLE
Basic setup for propagation node

### DISCUSSION THREAD

**Discussion by F4HVO** [markqvist/Reticulum#192] [Help & Questions] [answered]
Hi, I am starting to experiment with this technology and I am hitting a difficulty here.
I am running two nodes.
- Node 1 : windows computer plugged to a lora radio, Android phone running SideBand
- Node 2 : linux computer plugged to a lora radio and running lxmd and nomadenet

I can send direct messages between nodes 1 and 2. Now I would like to try to send a message fromNode 1 to a user that is not connected to the network. I would like Node 2 to 'buffer' the message and to forward it to the destination user when it connects to this node (or to any node of the network).

I was able to configure Node 2 to run lxmd and it now appears on SideBand (Node 1)  as a propagation node. I still can send direct messages to the user on Node 2, also I can send propagated messages from SideBand to the user on Node 2.

However, if the user is not connected to Node 2, user on Node 1 can send a propagated message, it appears as 'On Propagation Net' in SideBand but it stays in this state forever (even if the destination user connects to Node 2 using nomadenet.

**Comment by markqvist** [comment, marked as answer]
The primary reason this is happening is, that Sideband and NomadNet handles the use of Propagation Nodes slightly differently:

- The [Sideband](https://unsigned.io/sideband/) program will automatically use the closest propagation node it hears an annouce from, unless the user has manually chose a specific propagation node to be used. In this case, Sideband will *always* use that specific node, and that node *only*.
- The [Nomad Network](https://unsigned.io/nomadnet/) program will *not* automatically select a propagation node. Instead, Nomad Network handles propagation nodes like so:
    - In the list of saved nodes in the program, any saved node can be marked as either *Untrusted*, *Unknown* or *Trused*. Nomadnet will use the closest node marked as *Trusted* as the default propagation node
    - **Or** if the user has manually set a propagation node, it will use *that specific* node.
    - If the Nomad Network instance is itself set up as hosting a propagation node, it will automatically and periodically sync with all other reachable propagation nodes on the network, and messages intended for itself will of course be delivered to the Conversations view when they arrive
    - You should be able to see an announce from your `lxmd`-based propagation node in the announce stream of Nomad Network. If you mark that announce and press enter, you can select `< Use as default >`, which will make the program start using that node specifically.

The methodology for `nomadnet` is a bit more convoluted, and I admit, a bit opaque currently, but it was designed like that to allow much greater flexibility, whereas Sideband is focused more on "just working" out of the box, while still allowing basic configurability.

The user interaction for all of this in `nomadnet` could probably be cleaned up and improved, but I hope at least the above makes sense, and that it makes it more transparent to see how everything fits together.

**Comment by markqvist** [comment]
Also, out of curiosity, would you mind sharing what your Windows/Sideband setup is? Is it Sideband running in an Android emulator on Windows, but still being able to access the LoRa radio?

I'm pretty curious how this is set up :)

**Comment by F4HVO** [comment]
Okay thank you very much.
First, I was not 'syncing' after starting nomadenet, I needed to not open a conversation (or close it) and then hit ctrl+r. Also, I tried using SB on linux and it worked as expected (I also had to do a manual sync). I do not know if there is an option to do a sync on program start :)
Also I do not know how to 'save nodes' in nomadenet. I always see 'Currently no nodes are known' and in Local node info I see 'This instance is not hosting a node' which is confusing to me.

However, some messages still appear as 'On Propagation Net' (from the sending side, in SB) although I have received them. Since I have done a lot of tinkering I do not know yet how to reproduce this behaviour. It must be a variation of the pattern described previously.

About my setup, sorry for the misleading description, I used a windows pc to flash the firmware and to set the rnode in bluetooth pairing mode (under wsl 1). Then I used my phone with the SideBand app to send messages. No fancy emulation there sorry.

**Reply by markqvist** [reply]
> About my setup, sorry for the misleading description, I used a windows pc to flash the firmware and to set the rnode in bluetooth pairing mode (under wsl 1). Then I used my phone with the SideBand app to send messages. No fancy emulation there sorry.

Ah, no worries! That makes sense! Really good to hear that the flashing with `rnodeconf` worked on Windows too, I was unsure of whether that would be the case or not. Did you use WSL, or was it possible to run it in the plain old Windows command prompt?

> However, some messages still appear as 'On Propagation Net' (from the sending side, in SB) although I have received them. Since I have done a lot of tinkering I do not know yet how to reproduce this behaviour. It must be a variation of the pattern described previously.

When using propagation nodes, no message delivery notifications will be sent in return once the messages are received by the intended recipient. Currently, this is intentional, and done for privacy purposes. The functionality will be added later, but it needs to be done very carefully to protect user privacy and avoid metadata leaks.

So in short, you will currently only get real-time message delivery notifications for direct-mode deliveries.

> Also I do not know how to 'save nodes' in nomadenet. I always see 'Currently no nodes are known' and in Local node info I see 'This instance is not hosting a node' which is confusing to me.

Any node that appears in the "Announce Stream" list of the program can be saved by marking in the list, and hitting enter (or just clicking on it), and then selecting the `< Save >` button. You can use `Ctrl + L` in the `[ Network ]` section of the program to switch the list between the announce stream and the saved nodes list.

You can also save nodes you have open in the page browser by hitting `Ctrl + S`.

> I do not know if there is an option to do a sync on program start :)

There is, you can configure automatic periodic syncing for both Sideband and nomadnet in the settings/config of each program. The sync frequency is persistent, so if you start up your program, it will sync within a minute or two if it is time to do so, according to the interval (even if the program was closed in the meantime).

>  I needed to not open a conversation (or close it) and then hit ctrl+r.

It should be possible to get the sync dialog just by hitting `Ctrl + R`, it should not be necessarry to open and close conversations. You must have your cursor in the list or conversation pane though, the keyboard shortcuts are individual to the menu bar, so if your cursor is there, it wont work.

**Comment by F4HVO** [comment]
Fantastic ! Thank you.

> Did you use WSL, or was it possible to run it in the plain old Windows command prompt?

Yes I used the WSL. I tried windows command prompt but could not install `rnodeconf`.

`pip install rns`

![image](https://user-images.githubusercontent.com/54859526/215179979-b59b4c52-9630-44df-9191-1c4e6254975c.png)

**Reply by markqvist** [reply]
Nice, thanks for the info. Good to know for others who want to try it on Windows!
